internal/node/launcher.sh

```diff
@@ -257,24 +252,80 @@ if [[ "$PWD" == *"/bazel-out/"* ]]; then
-    # If BAZEL_NODE_MODULES_ROOT is set and it is not , add it to the list of bazel patch guards
-    # Also, add the external/${BAZEL_NODE_MODULES_ROOT} which is the correct path under execroot
-    # and under runfiles it is the legacy external runfiles path
-    export BAZEL_PATCH_GUARDS="${BAZEL_PATCH_GUARDS},${BAZEL_PATCH_ROOT}/${BAZEL_NODE_MODULES_ROOT},${PWD}/external/${BAZEL_NODE_MODULES_ROOT}"
+
+# Tell the node_patches_script that programs should not escape the execroot
+export BAZEL_PATCH_ROOTS="${EXECROOT}"
+# Set all bazel managed node_modules directories as guarded so no symlinks may
+# escape and no symlinks may enter.
+# We always guard against the root node_modules where 1st party deps go.
+# (e.g., /private/.../execroot/build_bazel_rules_nodejs/node_modules)
+export BAZEL_PATCH_ROOTS="${BAZEL_PATCH_ROOTS},${EXECROOT}/node_modules"
+if [[ "${RUNFILES_ROOT}" ]]; then
+  # If in runfiles, guard the runfiles root itself
+  export BAZEL_PATCH_ROOTS="${BAZEL_PATCH_ROOTS},${RUNFILES_ROOT}"
+  # If in runfiles guard the node_modules location in runfiles as well
+  # (e.g., /private/.../execroot/build_bazel_rules_nodejs/bazel-out/darwin-fastbuild/bin/internal/linker/test/multi_linker/test.sh.runfiles/build_bazel_rules_nodejs/node_modules)
+  export BAZEL_PATCH_ROOTS="${BAZEL_PATCH_ROOTS},${RUNFILES_ROOT}/${BAZEL_WORKSPACE}/node_modules"
+fi
+if [[ -n "${BAZEL_NODE_MODULES_ROOTS:-}" ]]; then
+  # BAZEL_NODE_MODULES_ROOTS is in the format "<path>:<workspace>,<path>:<workspace>"
+  # (e.g., "internal/linker/test:npm_internal_linker_test,:npm")
+  if [[ -n "${VERBOSE_LOGS:-}" ]]; then
+    echo "BAZEL_NODE_MODULES_ROOTS=${BAZEL_NODE_MODULES_ROOTS}" >&2
   fi
+  OLDIFS="${IFS}"
+  IFS=","
+  roots=(${BAZEL_NODE_MODULES_ROOTS})
+  IFS="${OLDIFS}"
+  for root in "${roots[@]}"; do
+    if [[ "${root}" ]]; then
+      OLDIFS="${IFS}"
+      IFS=":"
+      root_pair=(${root})
+      IFS="${OLDIFS}"
+      root_path="${root_pair[0]}"
+      root_workspace="${root_pair[1]}"
+      if [[ "${root_path}" ]]; then
+        # Guard non-root node_modules as well
+        # (e.g., /private/.../execroot/build_bazel_rules_nodejs/internal/linker/test/node_modules)
+        export BAZEL_PATCH_ROOTS="${BAZEL_PATCH_ROOTS},${EXECROOT}/${root_path}/node_modules"
+        if [[ "${RUNFILES_ROOT}" ]]; then
+          # If in runfiles guard the node_modules location in runfiles as well
+          # (e.g., /private/.../execroot/build_bazel_rules_nodejs/bazel-out/darwin-fastbuild/bin/internal/linker/test/multi_linker/test.sh.runfiles/build_bazel_rules_nodejs/internal/linker/test/node_modules)
+          export BAZEL_PATCH_ROOTS="${BAZEL_PATCH_ROOTS},${RUNFILES_ROOT}/${BAZEL_WORKSPACE}/${root_path}/node_modules"
+        fi
+      fi
+      # TODO: the following guards on the external workspaces may not be necessary and could be removed in the future with care
+      if [[ "${root_workspace}" ]] && [[ "${root_workspace}" != "${BAZEL_WORKSPACE}" ]]; then
```

Comment:
ideally feels like we should make an sh_test for this script, it has a lot of conditional logic and maybe amenable to unit test

Reply:
Seems like it has grown complicated enough to warrant one yeah. I'll make an issue and throw in in the 4.0 milestone so we don't forget about it